src/topology/fundamental_groupoid.lean

```diff
@@ -0,0 +1,458 @@
+{ to_fun := λ p, f.1 ⟨min (min (p.2.1+p.2.1) ((1-p.2.1)+(1-p.2.1))) (1-p.1.1), le_min (le_min
+        (add_nonneg p.2.2.1 p.2.2.1)
+        (add_nonneg (sub_nonneg.2 p.2.2.2) (sub_nonneg.2 p.2.2.2))) $
+      sub_nonneg_of_le p.1.2.2,
+    le_trans (min_le_right _ _) (sub_le_self _ p.1.2.1)⟩,
+  continuous := f.2.comp (continuous_induced_rng $
+    continuous_min (continuous_min
+      (continuous_add (continuous_induced_dom.comp continuous_snd) $
+        continuous_induced_dom.comp continuous_snd)
+      (continuous_add (continuous_sub continuous_const (continuous_induced_dom.comp continuous_snd))
+        (continuous_sub continuous_const $ continuous_induced_dom.comp continuous_snd)))
+      (continuous_sub continuous_const $ continuous_induced_dom.comp continuous_fst)),
+  initial := λ s, by simp only [set.Icc_zero_val, add_zero, sub_zero,
+      min_eq_left (add_nonneg (zero_le_one : (0:ℝ)≤1) zero_le_one),
+      min_eq_left (sub_nonneg_of_le s.2.2)]; convert f.3,
+  final := λ s, by simp only [set.Icc_one_val, sub_self, add_zero,
+      min_eq_right (add_nonneg (zero_le_one : (0:ℝ)≤1) zero_le_one),
+      min_eq_left (sub_nonneg_of_le s.2.2)]; convert f.3,
+  start := λ t, or.cases_on (le_total t.1 0.5)
+    (λ ht, by rw path.comp_apply_le _ _ _ ht; congr' 2; simp only [set.Icc_zero_val, sub_zero];
+      exact (min_eq_left $ le_trans (min_le_left _ _) $
+          le_trans (add_le_add ht ht) $ by rw add_halves).trans
+      (min_eq_left $ add_le_add (le_sub_iff_add_le.2 $
+          le_trans (add_le_add ht ht) $ by rw add_halves) $
+        le_sub_iff_add_le.2 $ le_trans (add_le_add ht ht) $ by rw add_halves))
+    (λ ht, by rw [path.comp_apply_ge _ _ _ ht, path.inv_apply]; congr' 2;
+      simp only [set.Icc_zero_val, sub_zero];
+      rw [← sub_add, sub_add_eq_add_sub 1 (t.1 + t.1), add_sub_comm]; exact
+      (min_eq_left $ le_trans (min_le_right _ _) $
+        le_trans (add_le_add (sub_le_sub_left ht _) (sub_le_sub_left ht _)) $
+        by rw [sub_half, add_halves]).trans
+      (min_eq_right $ add_le_add (sub_le_iff_le_add.2 $
+        le_trans (by rw add_halves) (add_le_add ht ht)) $
+        sub_le_iff_le_add.2 $ le_trans (by rw add_halves) (add_le_add ht ht))),
+  finish := λ t, by simp only [set.Icc_one_val, sub_self, path.id_apply];
+    refine eq.trans _ f.3; congr' 2; exact
+    min_eq_right (le_min (add_nonneg t.2.1 t.2.1) $
+    add_nonneg (sub_nonneg.2 t.2.2) $ sub_nonneg.2 t.2.2) }
+
+noncomputable def inv_comp (f : path x y) : homotopy (f.inv.comp f) (path.id y) :=
+by convert comp_inv f.inv; rw f.inv_inv
+
+noncomputable def id_comp (f : path x y) : homotopy ((path.id x).comp f) f :=
+by convert (comp_id f.inv).inv; simp only [inv_inv, path.inv_comp, path.inv_inv, path.inv_id]
+
+noncomputable def assoc (f : path w x) (g : path x y) (h : path y z) :
+  homotopy ((f.comp g).comp h) (f.comp (g.comp h)) :=
+{ to_fun := λ p, if 4 * p.2.1 / (1 + p.1.1) ≤ 1
```

Comment:
In math you ought to somehow avoid being able to do any computation at all here. Both sides are some continuous reparameterization of the function defined on [0, 3] by f, then g, then h, and whatever those reparameterizations [0, 1] -> [0, 3] are, you can linearly interpolate between them to define a homotopy which is constant at the endpoints. I'm not sure how easy it is to express that to Lean though (and this is the point where I gave up on this kind of approach when I tried this).

Same comments, but doable for sure, apply to `comp_id` and `comp_inv`. And even for definitions like `symm`, maybe it's better to define them in terms of what I called `congr_right` (or possibly `congr_left`, Scott has caused me to lose all sense of direction). If you have access to Switzer's book Algebraic Topology I think this stuff is explained properly in Chapter 2 there--all these basic constructions are induced by structure on the unit interval (or more generally, on a cylinder functor, as in https://github.com/rwbarton/lean-homotopy-theory/blob/lean-3.4.2/src/homotopy_theory/formal/cylinder/definitions.lean).